libnvme/ctrl.c

```diff
@@ -0,0 +1,439 @@
+	}
+
+	if (size > 0)
+		memcpy(rc_str, data, size);
+
+	errno = 0;
+	return rc_str;
+}
+
+int _nvme_ctrl_new(const char *nvme_ctrl_dev_path, struct nvme_ctrl **cnt,
+		   char *err_msg)
+{
+	int rc = NVME_OK;
+	int fd = -1;
+
+	assert(nvme_ctrl_dev_path != NULL);
+	assert(cnt != NULL);
+
+	*cnt = (struct nvme_ctrl *) calloc(1, sizeof(struct nvme_ctrl));
+	_alloc_null_check(err_msg, *cnt, rc, out);
+
+	(*cnt)->dev_path = strdup(nvme_ctrl_dev_path);
+
+	_alloc_null_check(err_msg, (*cnt)->dev_path, rc, out);
+
+	_good(_nvme_ioctl_fd_open(nvme_ctrl_dev_path, &fd, err_msg), rc, out);
+
+	_good(_nvme_ioctl_identify(fd, (uint8_t *) &((*cnt)->raw_data),
+				   _NVME_ADMIN_CMD_IDENTIFY_CNS_ALL_CTRL,
+				   0,
+				   err_msg),
+	      rc, out);
+
+out:
+	if (fd >= 0)
+		close(fd);
+	if (rc != NVME_OK) {
+		nvme_ctrl_free(*cnt);
+		*cnt = NULL;
+	}
+	return rc;
+}
+
+
+int nvme_ctrl_get_by_fd(int fd, struct nvme_ctrl **cnt, const char **err_msg)
+{
+	char err_msg_buff[_NVME_ERR_MSG_BUFF_LEN];
+	int rc = NVME_OK;
+	char *tmp_err_msg = NULL;
+
+	assert(cnt != NULL);
+
+	_nvme_err_msg_clear(err_msg_buff);
+	*cnt = NULL;
+	if (err_msg != NULL)
+		*err_msg = NULL;
+
+	*cnt = (struct nvme_ctrl *) calloc(1, sizeof(struct nvme_ctrl));
+	_alloc_null_check(err_msg_buff, *cnt, rc, out);
+
+	_good(_nvme_ioctl_identify(fd, (uint8_t *) &((*cnt)->raw_data),
+				   _NVME_ADMIN_CMD_IDENTIFY_CNS_ALL_CTRL,
+				   0,
+				   err_msg_buff),
+	      rc, out);
+
+out:
+	if (rc != NVME_OK) {
+		nvme_ctrl_free(*cnt);
+		*cnt = NULL;
+		if (err_msg != NULL) {
+			tmp_err_msg = (char *)
+				malloc(sizeof(char *) *
+				       _NVME_ERR_MSG_BUFF_LEN);
+			if (tmp_err_msg != NULL) {
+				memcpy(tmp_err_msg, err_msg_buff,
+				       _NVME_ERR_MSG_BUFF_LEN);
+				*err_msg = tmp_err_msg;
+			}
+		}
+	}
+	return rc;
+
+}
+
+void nvme_ctrl_free(struct nvme_ctrl *cnt)
+{
+	uint8_t i = 0;
+	if (cnt == NULL)
+		return;
+
+	for (i = 0; i <= nvme_ctrl_npss_get(cnt); ++i)
+		free(cnt->psds[i]);
+	free(cnt->psds);
+
+	free((void *) cnt->dev_path);
+	free((void *) cnt->strs.ver);
+	free((void *) cnt->strs.sn);
+	free((void *) cnt->strs.mn);
+	free((void *) cnt->strs.fr);
+	free((void *) cnt->strs.subnqn);
+	free(cnt);
+}
+
+const char *nvme_ctrl_dev_path_get(struct nvme_ctrl *cnt)
+{
+	assert(cnt != NULL);
+	errno = 0;
+	return cnt->dev_path;
+}
+
+const char *nvme_ctrl_ver_str_get(struct nvme_ctrl *cnt)
+{
+	char *rc_str = NULL;
+	struct _nvme_spec_ver *ver = NULL;
+
+	assert(cnt != NULL);
+	errno = 0;
+
+	if (cnt->strs.ver != NULL)
+		return cnt->strs.ver;
+
+	rc_str = (char *) calloc(_NVME_VER_STR_MAX_LEN, sizeof(char));
+
+	if (rc_str == NULL) {
+		errno = ENOMEM;
+		return NULL;
+	}
+
+	cnt->strs.ver = rc_str;
+
+	ver = (struct _nvme_spec_ver *) &(cnt->raw_data.ver);
+
+	snprintf(rc_str, _NVME_VER_STR_MAX_LEN,
+		 "%" PRIu16 ".%" PRIu8 ".%" PRIu8 "",
+		 le16toh(ver->mjr), ver->mnr, ver->ter);
+
+	if (strcmp(rc_str, "0.0.0") == 0)
+		snprintf(rc_str, _NVME_VER_STR_MAX_LEN, "1.0.0");
```

Comment:
Old NVMe standard 1.0 does not have version data defined, the `nvme_ctrl_ver_get()` will return 0 as raw data, for human readable data `nvme_ctrl_ver_str_get()`, we should translate that to version 1.0.0.